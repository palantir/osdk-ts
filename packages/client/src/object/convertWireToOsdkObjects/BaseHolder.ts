/*
 * Copyright 2025 Palantir Technologies, Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import type { ObjectOrInterfaceDefinition, ObjectSpecifier } from "@osdk/api";
import type { InterfaceHolder } from "./InterfaceHolder.js";
import type { UnderlyingOsdkObject } from "./InternalSymbols.js";
import type { ObjectHolder } from "./ObjectHolder.js";

/** @internal */

export interface BaseHolder {
  readonly [UnderlyingOsdkObject]: ObjectHolder;

  readonly $apiName: string;
  readonly $objectType: string;
  readonly $primaryKey: string | number;
  readonly $title: string | undefined;
  readonly $objectSpecifier: ObjectSpecifier<any>;

  readonly "$as": (
    newDef: string | ObjectOrInterfaceDefinition,
  ) => ObjectHolder | InterfaceHolder;

  readonly "$clone": (
    newProps?: Record<string, any>,
  ) => this;

  // [key: `$$${string}`]: any;
  // Unlike SimpleOsdkProperties, all of our remaining types are unknown as the full
  // union is basically `any` when you consider the above fields.
  [key: string]: unknown;
}
