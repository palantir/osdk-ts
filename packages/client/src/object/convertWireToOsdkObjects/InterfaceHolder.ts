/*
 * Copyright 2024 Palantir Technologies, Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import type { InterfaceDefinition } from "@osdk/api";
import type { FetchedObjectTypeDefinition } from "../../ontology/OntologyProvider.js";
import type { Osdk } from "../../OsdkObjectFrom.js";
import type {
  InterfaceDefRef,
  UnderlyingOsdkObject,
} from "./InternalSymbols.js";
import type { ObjectHolder } from "./ObjectHolder.js";

export interface InterfaceHolderOwnProps<
  Q extends FetchedObjectTypeDefinition<any, any>,
> {
  [UnderlyingOsdkObject]: Osdk<Q> & ObjectHolder<Q>;
  [InterfaceDefRef]: InterfaceDefinition<any>;
}

export interface InterfaceHolder<
  Q extends FetchedObjectTypeDefinition<any, any>,
> extends InterfaceHolderOwnProps<Q> {
}
