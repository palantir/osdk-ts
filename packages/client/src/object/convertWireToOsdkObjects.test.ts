/*
 * Copyright 2023 Palantir Technologies, Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import type { Attachment, Osdk } from "@osdk/client.api";
import {
  $ontologyRid,
  Employee,
  FooInterface,
  objectTypeWithAllPropertyTypes,
} from "@osdk/client.test.ontology";
import type { OntologyObjectV2 } from "@osdk/internal.foundry.core";
import { symbolClientContext } from "@osdk/shared.client";
import { createSharedClientContext } from "@osdk/shared.client.impl";
import { apiServer } from "@osdk/shared.test";
import {
  afterAll,
  beforeAll,
  describe,
  expect,
  expectTypeOf,
  it,
} from "vitest";
import type { Client } from "../Client.js";
import { createClient } from "../createClient.js";
import { createMinimalClient } from "../createMinimalClient.js";
import { convertWireToOsdkObjects } from "./convertWireToOsdkObjects.js";

describe("convertWireToOsdkObjects", () => {
  let client: Client;

  beforeAll(async () => {
    apiServer.listen();
    client = createClient(
      "https://stack.palantir.com",
      $ontologyRid,
      async () => "myAccessToken",
    );
  });

  afterAll(() => {
    apiServer.close();
  });

  it("configures properties correctly", async () => {
    const { data: [employee] } = await client(Employee).fetchPage();

    expect(Object.keys(employee).sort()).toEqual([
      "employeeId",
      "fullName",
      "office",
      "class",
      "startDate",
      "employeeStatus",
      "$apiName",
      "$objectType",
      "$primaryKey",
      "$title",
    ].sort());

    expect(Object.keys(employee.$as)).toEqual([]);
    expect(Object.keys(employee.$link)).toEqual([
      "peeps",
      "lead",
      "officeLink",
    ]);
  });

  it("reuses the object prototype across objects", async () => {
    const employees = await client(Employee).fetchPage();
    expect(employees.data.length).toBeGreaterThanOrEqual(2);
    const [a, b] = employees.data;

    expect(Object.getPrototypeOf(Object.getPrototypeOf(a))).toBe(
      Object.getPrototypeOf(Object.getPrototypeOf(b)),
    );
  });

  it("converts attachments as expected", async () => {
    const withValues = await client(
      objectTypeWithAllPropertyTypes,
    )
      .where({ id: 1 })
      .fetchPage();
    expect(withValues.data.length).toBeGreaterThanOrEqual(1);

    const { attachment, attachmentArray } = withValues.data[0];

    expectTypeOf(attachment).toMatchTypeOf<
      Attachment | undefined
    >;
    expect(attachment?.rid).toEqual(
      "ri.attachments.main.attachment.86016861-707f-4292-b258-6a7108915a75",
    );
    expect(Array.isArray(attachmentArray)).toBeTruthy();
    expectTypeOf(attachmentArray![0]).toMatchTypeOf<Attachment>;

    const withoutValues = await client(
      objectTypeWithAllPropertyTypes,
    ).where({ id: 2 }).fetchPage();

    const {
      attachment: emptyAttachment,
      attachmentArray: emptyAttachmentArray,
    } = withoutValues.data[0];
    expect(emptyAttachment).toBeUndefined();
    expect(emptyAttachmentArray).toBeUndefined();
  });

  it("works even with unknown apiNames", async () => {
    const clientCtx = createMinimalClient(
      { ontologyRid: $ontologyRid },
      "https://stack.palantir.com",
      async () => "myAccessToken",
    );
    createSharedClientContext(
      "https://stack.palantir.com",
      async () => "myAccessToken",
      "userAgent",
    );

    let object = {
      __apiName: Employee.apiName,
      __primaryKey: 0,
    } as const;
    const prototypeBefore = Object.getPrototypeOf(object);
    let object2 = await convertWireToOsdkObjects(
      clientCtx,
      [object],
      undefined,
      undefined,
      undefined,
      false,
    );
    const prototypeAfter = Object.getPrototypeOf(object2);

    expect(prototypeBefore).not.toBe(prototypeAfter);
  });

  it("updates interface when underlying changes", async () => {
    const clientCtx = createMinimalClient(
      { ontologyRid: $ontologyRid },
      "https://stack.palantir.com",
      async () => "myAccessToken",
    );

    let objectFromWire = {
      __apiName: "Employee" as const,
      __primaryKey: 0,
      __title: "Steve",
      fullName: "Steve",
      employeeId: "5",
    } satisfies OntologyObjectV2;

    const [obj] = (await convertWireToOsdkObjects(
      clientCtx,
      [objectFromWire],
      undefined,
    )) as unknown as Osdk<Employee>[];

    expect(obj.fullName).toEqual("Steve");
    expect(Object.keys(obj).sort()).toEqual([
      "$apiName",
      "$objectType",
      "$primaryKey",
      "$title",
      "employeeId",
      "fullName",
    ].sort());

    const objAsFoo = obj.$as(FooInterface);
    expect(objAsFoo).toMatchObject({
      fooSpt: obj.fullName,
      $apiName: FooInterface.apiName,
      $primaryKey: obj.$primaryKey,
      $objectType: obj.$objectType,
      $title: obj.$title,
    });

    console.log(obj);
    console.log(objAsFoo);

    (obj as any).$updateInternalValues({
      fullName: "Bob",
    });
    expect(obj.fullName).toEqual("Bob");
    expect(objAsFoo.fooSpt).toEqual(obj.fullName);

    expect(Object.keys(objAsFoo).sort()).toEqual([
      "$apiName",
      "$objectType",
      "$primaryKey",
      "$title",
      "fooSpt",
    ].sort());

    expect(obj).toBe(objAsFoo.$as(Employee));
    expect(objAsFoo).toBe(obj.$as(FooInterface));
  });

  it("reconstitutes interfaces properly without rid", async () => {
    const clientCtx = createMinimalClient(
      { ontologyRid: $ontologyRid },
      "https://stack.palantir.com",
      async () => "myAccessToken",
    );

    let objectFromWire = {
      __apiName: "Employee" as const,
      __primaryKey: 0,
      __title: "Steve",
      fooSpt: "Steve",
    } satisfies OntologyObjectV2;

    const [objAsFoo] = (await convertWireToOsdkObjects(
      clientCtx,
      [objectFromWire],
      FooInterface.apiName,
    )) as unknown as Osdk<FooInterface>[];

    expect(objAsFoo).toMatchInlineSnapshot(`
      {
        "$apiName": "FooInterface",
        "$objectType": "Employee",
        "$primaryKey": 0,
        "$title": "Steve",
        "fooSpt": "Steve",
      }
    `);

    const obj = objAsFoo.$as(Employee);
    expect(obj.fullName).toEqual("Steve");

    expect(obj).toMatchInlineSnapshot(`
      {
        "$apiName": "Employee",
        "$objectType": "Employee",
        "$primaryKey": 0,
        "$title": "Steve",
        "employeeId": 0,
        "fullName": "Steve",
      }
    `);
  });

  it("reconstitutes interfaces properly with rid", async () => {
    const clientCtx = createMinimalClient(
      { ontologyRid: $ontologyRid },
      "https://stack.palantir.com",
      async () => "myAccessToken",
    );

    let objectFromWire = {
      __apiName: "Employee" as const,
      __primaryKey: 0,
      __title: "Steve",
      __rid: "hiMom",
      fooSpt: "Steve",
    } satisfies OntologyObjectV2;

    const [objAsFoo] = (await convertWireToOsdkObjects(
      clientCtx,
      [objectFromWire],
      FooInterface.apiName,
    )) as unknown as Osdk<FooInterface, "$rid" | "$all">[];

    expect(objAsFoo).toMatchInlineSnapshot(`
      {
        "$apiName": "FooInterface",
        "$objectType": "Employee",
        "$primaryKey": 0,
        "$rid": "hiMom",
        "$title": "Steve",
        "fooSpt": "Steve",
      }
    `);
    expect(objAsFoo.$rid).toEqual("hiMom");

    const obj = objAsFoo.$as(Employee);
    expect(obj.fullName).toEqual("Steve");

    expect(obj).toMatchInlineSnapshot(`
      {
        "$apiName": "Employee",
        "$objectType": "Employee",
        "$primaryKey": 0,
        "$rid": "hiMom",
        "$title": "Steve",
        "employeeId": 0,
        "fullName": "Steve",
      }
    `);
    expect(obj.$rid).toEqual("hiMom");
  });

  describe("selection keys", () => {
    it("throws when required is missing", async () => {
      let object = {
        __apiName: "Employee",
        __primaryKey: 0,
      } as const;

      await expect(() =>
        convertWireToOsdkObjects(
          client[symbolClientContext],
          [object],
          undefined,
          undefined,
          ["employeeId"],
          "throw",
        )
      ).rejects.toThrowErrorMatchingInlineSnapshot(
        `[Error: Unable to safely convert objects as some non nullable properties are null]`,
      );
    });

    it("does not throw when optional is missing", async () => {
      let object = {
        __apiName: "Employee",
        __primaryKey: 0,
      } as const;

      await expect(
        convertWireToOsdkObjects(
          client[symbolClientContext],
          [object],
          undefined,
          undefined,
          ["fullName"],
          "throw",
        ),
      ).resolves.to.not.toBeUndefined();
    });

    it("filters when it should", async () => {
      const object = {
        __apiName: "Employee",
        __primaryKey: 0,
      } as const;

      const result = await convertWireToOsdkObjects(
        client[symbolClientContext],
        [object],
        undefined,
        undefined,
        ["employeeId"],
        "drop",
      );

      expect(result.length).toBe(0);
    });

    it("does not filter when it shouldn't", async () => {
      const object = {
        __apiName: "Employee",
        __primaryKey: 0,
      } as const;

      const result = await convertWireToOsdkObjects(
        client[symbolClientContext],
        [object],
        undefined,
        undefined,
        ["fullName"],
        "drop",
      );

      expect(result.length).toBe(1);
    });
  });

  describe("without selection keys", () => {
    it("throws when required is missing", async () => {
      let object = {
        __apiName: "Employee",
        __primaryKey: 0,
      } as const;

      await expect(() =>
        convertWireToOsdkObjects(
          client[symbolClientContext],
          [object],
          undefined,
          undefined,
          undefined,
          "throw",
        )
      ).rejects.toThrowErrorMatchingInlineSnapshot(
        `[Error: Unable to safely convert objects as some non nullable properties are null]`,
      );
    });

    it("does not throw when required is present", async () => {
      let object = {
        __apiName: "Employee",
        __primaryKey: 0,
        "employeeId": 0,
      } as const;

      await expect(
        convertWireToOsdkObjects(
          client[symbolClientContext],
          [object],
          undefined,
          undefined,
          undefined,
          "throw",
        ),
      ).resolves.to.not.toBeUndefined();
    });

    it("filters when it should", async () => {
      const object = {
        __apiName: "Employee",
        __primaryKey: 0,
      } as const;

      const result = await convertWireToOsdkObjects(
        client[symbolClientContext],
        [object],
        undefined,
        undefined,
        undefined,
        "drop",
      );

      expect(result.length).toBe(0);
    });

    it("does not filter when it shouldn't", async () => {
      const object = {
        __apiName: "Employee",
        __primaryKey: 0,
        "employeeId": 0,
      } as const;

      const result = await convertWireToOsdkObjects(
        client[symbolClientContext],
        [object],
        undefined,
        undefined,
        undefined,
        "drop",
      );

      expect(result.length).toBe(1);
    });
  });

  it("behaves correctly when converting", async () => {
    const object = {
      __apiName: "Employee",
      __primaryKey: 0,
      fooSpt: "hi",
    } as const;

    const result = await convertWireToOsdkObjects(
      client[symbolClientContext],
      [object],
      "FooInterface",
      undefined,
      ["fooSpt"],
      "drop",
    );

    expect(result.length).toBe(1);
  });
});
