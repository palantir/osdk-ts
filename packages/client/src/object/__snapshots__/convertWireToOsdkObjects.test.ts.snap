// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`convertWireToOsdkObjects > reconstitutes interfaces properly with rid - new 2`] = `
{
  "$apiName": "Employee",
  "$metadata": {
    "apiName": "Employee",
    "description": "A full-time or part-time 

 employee of our firm",
    "displayName": "Employee",
    "icon": {
      "color": "blue",
      "name": "person",
      "type": "blueprint",
    },
    "implements": [
      "FooInterface",
    ],
    "interfaceMap": {
      "FooInterface": {
        "fooSpt": "fullName",
      },
    },
    "inverseInterfaceMap": {
      "FooInterface": {
        "fullName": "fooSpt",
      },
    },
    "links": {
      "lead": {
        "multiplicity": false,
        "targetType": "Employee",
      },
      "officeLink": {
        "multiplicity": false,
        "targetType": "Office",
      },
      "peeps": {
        "multiplicity": true,
        "targetType": "Employee",
      },
    },
    "pluralDisplayName": "Employees",
    "primaryKeyApiName": "employeeId",
    "primaryKeyType": "integer",
    "properties": {
      "class": {
        "description": "",
        "displayName": undefined,
        "multiplicity": false,
        "nullable": true,
        "type": "string",
      },
      "employeeId": {
        "description": undefined,
        "displayName": undefined,
        "multiplicity": false,
        "nullable": false,
        "type": "integer",
      },
      "employeeLocation": {
        "description": "Geotime series reference of the location of the employee",
        "displayName": undefined,
        "multiplicity": false,
        "nullable": true,
        "type": "geotimeSeriesReference",
      },
      "employeeSensor": {
        "description": "TimeSeries sensor of the status of the employee",
        "displayName": undefined,
        "multiplicity": false,
        "nullable": true,
        "type": "sensorTimeseries",
      },
      "employeeStatus": {
        "description": "TimeSeries of the status of the employee",
        "displayName": undefined,
        "multiplicity": false,
        "nullable": true,
        "type": "stringTimeseries",
      },
      "fullName": {
        "description": undefined,
        "displayName": undefined,
        "multiplicity": false,
        "nullable": true,
        "type": "string",
      },
      "office": {
        "description": "The unique "ID" of the employee's \\"primary\\" assigned office.
 This is some more text.",
        "displayName": undefined,
        "multiplicity": false,
        "nullable": true,
        "type": "string",
      },
      "startDate": {
        "description": "The date the employee was hired (most recently, if they were re-hired)",
        "displayName": undefined,
        "multiplicity": false,
        "nullable": true,
        "type": "datetime",
      },
    },
    "rid": "ri.ontology.main.object-type.401ac022-89eb-4591-8b7e-0a912b9efb44",
    "status": "ACTIVE",
    "titleProperty": "fullName",
    "type": "object",
    "visibility": "NORMAL",
    Symbol(InterfaceDefinitions): {
      "FooInterface": {
        "def": {
          "apiName": "FooInterface",
          "description": "Interface for Foo",
          "displayName": "Foo Interface",
          "implementedBy": [
            "Employee",
            "Person",
          ],
          "implements": [],
          "links": {},
          "properties": {
            "fooSpt": {
              "description": "A foo",
              "displayName": "Foo",
              "multiplicity": false,
              "nullable": true,
              "type": "string",
            },
          },
          "rid": "ri.interface.main.interface.1",
          "type": "interface",
        },
        "handler": undefined,
      },
    },
  },
  "$objectSpecifier": "Employee:0",
  "$objectType": "Employee",
  "$primaryKey": 0,
  "$rid": "hiMom",
  "$title": "Steve",
  "employeeId": 0,
  "fullName": "Steve",
}
`;

exports[`convertWireToOsdkObjects > reconstitutes interfaces properly with rid 2`] = `
{
  "$apiName": "Employee",
  "$metadata": {
    "apiName": "Employee",
    "description": "A full-time or part-time 

 employee of our firm",
    "displayName": "Employee",
    "icon": {
      "color": "blue",
      "name": "person",
      "type": "blueprint",
    },
    "implements": [
      "FooInterface",
    ],
    "interfaceMap": {
      "FooInterface": {
        "fooSpt": "fullName",
      },
    },
    "inverseInterfaceMap": {
      "FooInterface": {
        "fullName": "fooSpt",
      },
    },
    "links": {
      "lead": {
        "multiplicity": false,
        "targetType": "Employee",
      },
      "officeLink": {
        "multiplicity": false,
        "targetType": "Office",
      },
      "peeps": {
        "multiplicity": true,
        "targetType": "Employee",
      },
    },
    "pluralDisplayName": "Employees",
    "primaryKeyApiName": "employeeId",
    "primaryKeyType": "integer",
    "properties": {
      "class": {
        "description": "",
        "displayName": undefined,
        "multiplicity": false,
        "nullable": true,
        "type": "string",
      },
      "employeeId": {
        "description": undefined,
        "displayName": undefined,
        "multiplicity": false,
        "nullable": false,
        "type": "integer",
      },
      "employeeLocation": {
        "description": "Geotime series reference of the location of the employee",
        "displayName": undefined,
        "multiplicity": false,
        "nullable": true,
        "type": "geotimeSeriesReference",
      },
      "employeeSensor": {
        "description": "TimeSeries sensor of the status of the employee",
        "displayName": undefined,
        "multiplicity": false,
        "nullable": true,
        "type": "sensorTimeseries",
      },
      "employeeStatus": {
        "description": "TimeSeries of the status of the employee",
        "displayName": undefined,
        "multiplicity": false,
        "nullable": true,
        "type": "stringTimeseries",
      },
      "fullName": {
        "description": undefined,
        "displayName": undefined,
        "multiplicity": false,
        "nullable": true,
        "type": "string",
      },
      "office": {
        "description": "The unique "ID" of the employee's \\"primary\\" assigned office.
 This is some more text.",
        "displayName": undefined,
        "multiplicity": false,
        "nullable": true,
        "type": "string",
      },
      "startDate": {
        "description": "The date the employee was hired (most recently, if they were re-hired)",
        "displayName": undefined,
        "multiplicity": false,
        "nullable": true,
        "type": "datetime",
      },
    },
    "rid": "ri.ontology.main.object-type.401ac022-89eb-4591-8b7e-0a912b9efb44",
    "status": "ACTIVE",
    "titleProperty": "fullName",
    "type": "object",
    "visibility": "NORMAL",
    Symbol(InterfaceDefinitions): {
      "FooInterface": {
        "def": {
          "apiName": "FooInterface",
          "description": "Interface for Foo",
          "displayName": "Foo Interface",
          "implementedBy": [
            "Employee",
            "Person",
          ],
          "implements": [],
          "links": {},
          "properties": {
            "fooSpt": {
              "description": "A foo",
              "displayName": "Foo",
              "multiplicity": false,
              "nullable": true,
              "type": "string",
            },
          },
          "rid": "ri.interface.main.interface.1",
          "type": "interface",
        },
        "handler": undefined,
      },
    },
  },
  "$objectSpecifier": "Employee:0",
  "$objectType": "Employee",
  "$primaryKey": 0,
  "$rid": "hiMom",
  "$title": "Steve",
  "employeeId": 0,
  "fullName": "Steve",
}
`;

exports[`convertWireToOsdkObjects > reconstitutes interfaces properly without rid - new 2`] = `
{
  "$apiName": "Employee",
  "$metadata": {
    "apiName": "Employee",
    "description": "A full-time or part-time 

 employee of our firm",
    "displayName": "Employee",
    "icon": {
      "color": "blue",
      "name": "person",
      "type": "blueprint",
    },
    "implements": [
      "FooInterface",
    ],
    "interfaceMap": {
      "FooInterface": {
        "fooSpt": "fullName",
      },
    },
    "inverseInterfaceMap": {
      "FooInterface": {
        "fullName": "fooSpt",
      },
    },
    "links": {
      "lead": {
        "multiplicity": false,
        "targetType": "Employee",
      },
      "officeLink": {
        "multiplicity": false,
        "targetType": "Office",
      },
      "peeps": {
        "multiplicity": true,
        "targetType": "Employee",
      },
    },
    "pluralDisplayName": "Employees",
    "primaryKeyApiName": "employeeId",
    "primaryKeyType": "integer",
    "properties": {
      "class": {
        "description": "",
        "displayName": undefined,
        "multiplicity": false,
        "nullable": true,
        "type": "string",
      },
      "employeeId": {
        "description": undefined,
        "displayName": undefined,
        "multiplicity": false,
        "nullable": false,
        "type": "integer",
      },
      "employeeLocation": {
        "description": "Geotime series reference of the location of the employee",
        "displayName": undefined,
        "multiplicity": false,
        "nullable": true,
        "type": "geotimeSeriesReference",
      },
      "employeeSensor": {
        "description": "TimeSeries sensor of the status of the employee",
        "displayName": undefined,
        "multiplicity": false,
        "nullable": true,
        "type": "sensorTimeseries",
      },
      "employeeStatus": {
        "description": "TimeSeries of the status of the employee",
        "displayName": undefined,
        "multiplicity": false,
        "nullable": true,
        "type": "stringTimeseries",
      },
      "fullName": {
        "description": undefined,
        "displayName": undefined,
        "multiplicity": false,
        "nullable": true,
        "type": "string",
      },
      "office": {
        "description": "The unique "ID" of the employee's \\"primary\\" assigned office.
 This is some more text.",
        "displayName": undefined,
        "multiplicity": false,
        "nullable": true,
        "type": "string",
      },
      "startDate": {
        "description": "The date the employee was hired (most recently, if they were re-hired)",
        "displayName": undefined,
        "multiplicity": false,
        "nullable": true,
        "type": "datetime",
      },
    },
    "rid": "ri.ontology.main.object-type.401ac022-89eb-4591-8b7e-0a912b9efb44",
    "status": "ACTIVE",
    "titleProperty": "fullName",
    "type": "object",
    "visibility": "NORMAL",
    Symbol(InterfaceDefinitions): {
      "FooInterface": {
        "def": {
          "apiName": "FooInterface",
          "description": "Interface for Foo",
          "displayName": "Foo Interface",
          "implementedBy": [
            "Employee",
            "Person",
          ],
          "implements": [],
          "links": {},
          "properties": {
            "fooSpt": {
              "description": "A foo",
              "displayName": "Foo",
              "multiplicity": false,
              "nullable": true,
              "type": "string",
            },
          },
          "rid": "ri.interface.main.interface.1",
          "type": "interface",
        },
        "handler": undefined,
      },
    },
  },
  "$objectSpecifier": "Employee:0",
  "$objectType": "Employee",
  "$primaryKey": 0,
  "$title": "Steve",
  "fullName": "Steve",
}
`;

exports[`convertWireToOsdkObjects > reconstitutes interfaces properly without rid - old 1`] = `
{
  "$apiName": "FooInterface",
  "$metadata": {
    "apiName": "FooInterface",
    "description": "Interface for Foo",
    "displayName": "Foo Interface",
    "implementedBy": [
      "Employee",
      "Person",
    ],
    "implements": [],
    "links": {},
    "properties": {
      "fooSpt": {
        "description": "A foo",
        "displayName": "Foo",
        "multiplicity": false,
        "nullable": true,
        "type": "string",
      },
    },
    "rid": "ri.interface.main.interface.1",
    "type": "interface",
  },
  "$objectSpecifier": "Employee:0",
  "$objectType": "Employee",
  "$primaryKey": 0,
  "$title": "Steve",
  "fooSpt": "Steve",
}
`;

exports[`convertWireToOsdkObjects > reconstitutes interfaces properly without rid - old 2`] = `
{
  "$apiName": "Employee",
  "$metadata": {
    "apiName": "Employee",
    "description": "A full-time or part-time 

 employee of our firm",
    "displayName": "Employee",
    "icon": {
      "color": "blue",
      "name": "person",
      "type": "blueprint",
    },
    "implements": [
      "FooInterface",
    ],
    "interfaceMap": {
      "FooInterface": {
        "fooSpt": "fullName",
      },
    },
    "inverseInterfaceMap": {
      "FooInterface": {
        "fullName": "fooSpt",
      },
    },
    "links": {
      "lead": {
        "multiplicity": false,
        "targetType": "Employee",
      },
      "officeLink": {
        "multiplicity": false,
        "targetType": "Office",
      },
      "peeps": {
        "multiplicity": true,
        "targetType": "Employee",
      },
    },
    "pluralDisplayName": "Employees",
    "primaryKeyApiName": "employeeId",
    "primaryKeyType": "integer",
    "properties": {
      "class": {
        "description": "",
        "displayName": undefined,
        "multiplicity": false,
        "nullable": true,
        "type": "string",
      },
      "employeeId": {
        "description": undefined,
        "displayName": undefined,
        "multiplicity": false,
        "nullable": false,
        "type": "integer",
      },
      "employeeLocation": {
        "description": "Geotime series reference of the location of the employee",
        "displayName": undefined,
        "multiplicity": false,
        "nullable": true,
        "type": "geotimeSeriesReference",
      },
      "employeeSensor": {
        "description": "TimeSeries sensor of the status of the employee",
        "displayName": undefined,
        "multiplicity": false,
        "nullable": true,
        "type": "sensorTimeseries",
      },
      "employeeStatus": {
        "description": "TimeSeries of the status of the employee",
        "displayName": undefined,
        "multiplicity": false,
        "nullable": true,
        "type": "stringTimeseries",
      },
      "fullName": {
        "description": undefined,
        "displayName": undefined,
        "multiplicity": false,
        "nullable": true,
        "type": "string",
      },
      "office": {
        "description": "The unique "ID" of the employee's \\"primary\\" assigned office.
 This is some more text.",
        "displayName": undefined,
        "multiplicity": false,
        "nullable": true,
        "type": "string",
      },
      "startDate": {
        "description": "The date the employee was hired (most recently, if they were re-hired)",
        "displayName": undefined,
        "multiplicity": false,
        "nullable": true,
        "type": "datetime",
      },
    },
    "rid": "ri.ontology.main.object-type.401ac022-89eb-4591-8b7e-0a912b9efb44",
    "status": "ACTIVE",
    "titleProperty": "fullName",
    "type": "object",
    "visibility": "NORMAL",
    Symbol(InterfaceDefinitions): {
      "FooInterface": {
        "def": {
          "apiName": "FooInterface",
          "description": "Interface for Foo",
          "displayName": "Foo Interface",
          "implementedBy": [
            "Employee",
            "Person",
          ],
          "implements": [],
          "links": {},
          "properties": {
            "fooSpt": {
              "description": "A foo",
              "displayName": "Foo",
              "multiplicity": false,
              "nullable": true,
              "type": "string",
            },
          },
          "rid": "ri.interface.main.interface.1",
          "type": "interface",
        },
        "handler": undefined,
      },
    },
  },
  "$objectSpecifier": "Employee:0",
  "$objectType": "Employee",
  "$primaryKey": 0,
  "$title": "Steve",
  "employeeId": 0,
  "fullName": "Steve",
}
`;

exports[`convertWireToOsdkObjects > stringifies properties on objects and interfaces correctly 1`] = `"{"employeeId":50030,"fullName":"John Doe","office":"NYC","class":"Red","startDate":"2019-01-01","employeeStatus":{},"employeeSensor":{},"employeeLocation":{},"$apiName":"Employee","$objectType":"Employee","$primaryKey":50030,"$objectSpecifier":"Employee:50030","$metadata":{"type":"object","apiName":"Employee","description":"A full-time or part-time \\n\\n employee of our firm","primaryKeyApiName":"employeeId","primaryKeyType":"integer","links":{"peeps":{"multiplicity":true,"targetType":"Employee"},"lead":{"multiplicity":false,"targetType":"Employee"},"officeLink":{"multiplicity":false,"targetType":"Office"}},"properties":{"employeeId":{"multiplicity":false,"type":"integer","nullable":false},"fullName":{"multiplicity":false,"type":"string","nullable":true},"office":{"multiplicity":false,"description":"The unique \\"ID\\" of the employee's \\\\\\"primary\\\\\\" assigned office.\\n This is some more text.","type":"string","nullable":true},"class":{"multiplicity":false,"description":"","type":"string","nullable":true},"startDate":{"multiplicity":false,"description":"The date the employee was hired (most recently, if they were re-hired)","type":"datetime","nullable":true},"employeeStatus":{"multiplicity":false,"description":"TimeSeries of the status of the employee","type":"stringTimeseries","nullable":true},"employeeSensor":{"multiplicity":false,"description":"TimeSeries sensor of the status of the employee","type":"sensorTimeseries","nullable":true},"employeeLocation":{"multiplicity":false,"description":"Geotime series reference of the location of the employee","type":"geotimeSeriesReference","nullable":true}},"implements":["FooInterface"],"interfaceMap":{"FooInterface":{"fooSpt":"fullName"}},"inverseInterfaceMap":{"FooInterface":{"fullName":"fooSpt"}},"icon":{"type":"blueprint","color":"blue","name":"person"},"titleProperty":"fullName","displayName":"Employee","pluralDisplayName":"Employees","status":"ACTIVE","rid":"ri.ontology.main.object-type.401ac022-89eb-4591-8b7e-0a912b9efb44","visibility":"NORMAL"}}"`;

exports[`convertWireToOsdkObjects > stringifies properties on objects and interfaces correctly 3`] = `"{"employeeId":50031,"fullName":"Jane Doe","office":"SEA","class":"Blue","startDate":"2012-02-12","employeeStatus":{},"employeeSensor":{},"employeeLocation":{},"$apiName":"Employee","$objectType":"Employee","$primaryKey":50031,"$title":"Jane Doe","$objectSpecifier":"Employee:50031","$metadata":{"type":"object","apiName":"Employee","description":"A full-time or part-time \\n\\n employee of our firm","primaryKeyApiName":"employeeId","primaryKeyType":"integer","links":{"peeps":{"multiplicity":true,"targetType":"Employee"},"lead":{"multiplicity":false,"targetType":"Employee"},"officeLink":{"multiplicity":false,"targetType":"Office"}},"properties":{"employeeId":{"multiplicity":false,"type":"integer","nullable":false},"fullName":{"multiplicity":false,"type":"string","nullable":true},"office":{"multiplicity":false,"description":"The unique \\"ID\\" of the employee's \\\\\\"primary\\\\\\" assigned office.\\n This is some more text.","type":"string","nullable":true},"class":{"multiplicity":false,"description":"","type":"string","nullable":true},"startDate":{"multiplicity":false,"description":"The date the employee was hired (most recently, if they were re-hired)","type":"datetime","nullable":true},"employeeStatus":{"multiplicity":false,"description":"TimeSeries of the status of the employee","type":"stringTimeseries","nullable":true},"employeeSensor":{"multiplicity":false,"description":"TimeSeries sensor of the status of the employee","type":"sensorTimeseries","nullable":true},"employeeLocation":{"multiplicity":false,"description":"Geotime series reference of the location of the employee","type":"geotimeSeriesReference","nullable":true}},"implements":["FooInterface"],"interfaceMap":{"FooInterface":{"fooSpt":"fullName"}},"inverseInterfaceMap":{"FooInterface":{"fullName":"fooSpt"}},"icon":{"type":"blueprint","color":"blue","name":"person"},"titleProperty":"fullName","displayName":"Employee","pluralDisplayName":"Employees","status":"ACTIVE","rid":"ri.ontology.main.object-type.401ac022-89eb-4591-8b7e-0a912b9efb44","visibility":"NORMAL"}}"`;
